### PR TITLE
Implements pbkdf2 for rc_crypto

### LINKED ISSUE #3190: Add pbkdf2 to rc_crypto
Extracted from [https://github.com/mozilla/application-services/pull/3187/files#r433409179](https://github.com/mozilla/application-services/pull/3187/files#r433409179)

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-1362)
┆Story Points: 5
┆Sprint: SYNC - end 2020-06-05

### PR DESCRIPTION
**Pull request by tarikeshaq**
### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `automation/all_tests.sh` runs to completion and produces no failures
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

fixes #3190 
Adds [pbkdf2](https://www.ietf.org/rfc/rfc2898.txt) to rc_crypto. To be used by #3187 and any future uses for extending passwords.

### CODE REVIEW

File: components/support/rc_crypto/src/pbkdf2.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+///```
+///
+/// # Erros
+///
+/// Could possibly return an error if the HMAC algorithm fails
+pub fn pbkdf2_generate(
+    passphrase: &[u8],
+    salt: &[u8],
+    iterations: u32,
+    key_len: u32,
```
Comment (rfk) [inline components/support/rc_crypto/src/pbkdf2.rs:33]: API nit: I think it would be more idiomatic to let the caller provide an output buffer `out: &mut [u8]` and have this function will it, rather than allocate a new buffer. See e.g. the API exposed by `hkdf.rs` in this same directory.

File: components/support/rc_crypto/src/pbkdf2.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            *curr_byte ^= u_curr.as_ref()[k];
+        }
+        u_prev = u_curr;
+    }
+    Ok(res)
+}
+
+fn to_32b_buf(n: u32) -> Vec<u8> {
+    let mut buf = [0; 4];
+    BigEndian::write_u32(&mut buf, n);
```
Comment (rfk) [inline components/support/rc_crypto/src/pbkdf2.rs:70]: Using [`u32::to_be_bytes`](https://doc.rust-lang.org/std/primitive.u32.html#method.to_be_bytes) might help avoid an external dependency here.

**Comment by rfk** [comment]
I should add: the algorithm is simple enough (and the risks of doing it wrong are low enough) that I have no objection to us shipping our own implementation, if NSS doesn't natively provide what we want or if it's a PITA to use.

**Comment by tarikeshaq** [comment]
Thanks again for the comments 😄 I'll look into NSS and see what I can figure out there!

**Comment by tarikeshaq** [comment]
So since I'm calling into NSS, quite a bit of unsafe code exists and since this my first time writing unsafe code (although it is just calling into NSS that's unsafe) I would love any pointers on what could change / any bad unsafe practices 😄  etc

Any advice on the best way to clean up the pointers would also be great! I'm considering just having a helper function that does it, but I was wondering if something like the stuff in nss/pk11/sym_key would be better (In fact I'm considering moving everything there similar to hkdf_expand, would that be the best approach)

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+// https://searchfox.org/mozilla-central/rev/8ccea36c4fb09412609fb738c722830d7098602b/dom/crypto/WebCryptoTask.cpp#2567
+
+pub fn nss_pbkdf2_generate(
+    password: &mut [u8],
+    salt: &mut [u8],
+    iterations: u32,
+    out: &mut [u8],
+) -> Result<()> {
+    ensure_nss_initialized();
+    // Todo: Maybe add support for different hashing algorithms?
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:21]: Agree, how would you feel about adding an `Algorithm` param for this function?

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        data: salt.as_mut_ptr(),
+        type_: 0,
+    };
+
+    let alg_id = unsafe {
+        nss_sys::PK11_CreatePBEV2AlgorithmID(
+            SECOidTag::SEC_OID_PKCS5_PBKDF2 as u32,
+            SECOidTag::SEC_OID_HMAC_SHA1 as u32,
+            oid_tag as u32,
+            // Should add error handling to prevent panic if (in the exteremely unlikely scenario) the cast can't be done
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:34]: Can't we use `try_from` here? If not we should do manual bounds checking.

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    // Todo: Maybe add support for different hashing algorithms?
+    let oid_tag = SECOidTag::SEC_OID_HMAC_SHA256;
+    let mut sec_salt = nss_sys::SECItem {
+        len: salt.len() as u32,
+        data: salt.as_mut_ptr(),
+        type_: 0,
+    };
+
+    let alg_id = unsafe {
+        nss_sys::PK11_CreatePBEV2AlgorithmID(
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:30]: You should wrap this call in a [scoped pointer wrapper](https://searchfox.org/mozilla-central/rev/7cadba1d8b8feaec4615f5bb98aac4b8a719793c/security/manager/ssl/ScopedNSSTypes.h#304-305). We do this using the `scoped_ptr!` macro, there's some examples in components/support/rc_crypto/nss/src/pk11/types.rs.

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            iterations as i32,
+            &mut sec_salt as *mut nss_sys::SECItem,
+        )
+    };
+
+    if alg_id.is_null() {
+        return Err(ErrorKind::InternalError.into());
+    }
+
+    let slot = unsafe { nss_sys::PK11_GetInternalSlot() };
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:45]: We have a safe wrapper for that `get_internal_slot`, can we use it?

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        unsafe { nss_sys::SECOID_DestroyAlgorithmID(alg_id, nss_sys::PR_TRUE) };
+        return Err(ErrorKind::InternalError.into());
+    }
+    let mut pw_item = nss_sys::SECItem {
+        len: password.len() as u32,
+        data: password.as_mut_ptr(),
+        type_: 0,
+    };
+
+    let sym_key = unsafe {
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:56]: I think you can use the `SymKey` wrapper struct here.

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                nss_sys::PK11_FreeSlot(slot);
+            };
+        }
+        if !sym_key.is_null() {
+            unsafe {
+                nss_sys::PK11_FreeSymKey(sym_key);
+            };
+        }
+        return Err(ErrorKind::InternalError.into());
+    }
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:84]: 😅

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    let data = unsafe {
+        let inner = (*key_data).data;
+        std::slice::from_raw_parts_mut(inner, out.len())
+    };
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:109]: Do you want `sec_item_as_slice`?

File: components/support/rc_crypto/nss/nss_sys/src/bindings/pk11pub.rs [hunk 10]
```diff
@@ -110,4 +110,20 @@ extern "C" {
...
+        salt: *mut SECItem,
+    ) -> *mut SECAlgorithmID;
+
+    pub fn PK11_PBEKeyGen(
+        slot: *mut PK11SlotInfo,
+        algid: *mut SECAlgorithmID,
+        pwitem: *mut SECItem,
+        faulty3DES: PRBool,
+        wincx: *mut c_void,
+    ) -> *mut PK11SymKey;
```
Comment (eoger) [inline components/support/rc_crypto/nss/nss_sys/src/bindings/pk11pub.rs:128]: Did you generate these with bindgen or hand-written? (in any case I assume ctest will catch a problem)

Comment (eoger) [reply components/support/rc_crypto/nss/nss_sys/src/bindings/pk11pub.rs:128]: I manually checked these bindings against the nss header files and these look correct 👍

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+use crate::error::*;
+use crate::util::ensure_nss_initialized;
+use nss_sys::SECOidTag;
+
+// ***** BASED ON THE FOLLOWING IMPLEMENTATION *****
+// https://github.com/sqlcipher/sqlcipher/blob/4a81bea61e1da6fec222d713852830f1fd01aed2/src/crypto_nss.c#L182-L232
+// Which I believe is based on how it is done on WebCrypto:
+// https://searchfox.org/mozilla-central/rev/8ccea36c4fb09412609fb738c722830d7098602b/dom/crypto/WebCryptoTask.cpp#2567
+
+pub fn nss_pbkdf2_generate(
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:14]: Nit: Can we call this `pbkdf2_key_derive`, `pbkdf2_keygen` or something else?

File: components/support/rc_crypto/nss/src/pbkdf2.rs (hunk 3 above)
Comment (tarikeshaq) [reply components/support/rc_crypto/nss/src/pbkdf2.rs]: Done! (Used HashAlgorithm from context, not sure if that's the right choice, but seemed handy)

File: components/support/rc_crypto/nss/src/pbkdf2.rs (hunk 5 above)
Comment (tarikeshaq) [reply components/support/rc_crypto/nss/src/pbkdf2.rs]: I had to add to the scoped_ptr! macro a second expression where the destructor has an argument (since the algorithm ID destructor requires an extra argument) wasn't sure if that is best, or if I should define a separate struct for AlgorithmID without using the macro.

Comment (eoger) [reply components/support/rc_crypto/nss/src/pbkdf2.rs]: `pk11_destroy_context_true` is an example where we send a param to the destructor

File: components/support/rc_crypto/nss/src/pk11/sym_key.rs [hunk 12]
```diff
@@ -91,3 +95,23 @@ pub(crate) fn import_sym_key(
         ))
     }
 }
+
+/// Safe wrapper around PK11_PBEKeyGen that
+/// de-allocates memory when the key goes out of
+/// scope.
+pub(crate) fn pbe_key_gen(
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pk11/sym_key.rs:102]: I think that's fine to inline into the caller, since there's only one use of this.

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    salt: &mut [u8],
+    iterations: u32,
+    hash_algorithm: HashAlgorithm,
+    out: &mut [u8],
+) -> Result<()> {
+    ensure_nss_initialized();
+    let oid_tag = match hash_algorithm {
+        HashAlgorithm::SHA256 => SECOidTag::SEC_OID_HMAC_SHA256 as u32,
+    };
+
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:33]: Very very nit: there's a ton of line breaks in this function which makes it bit harder to reason about. Can we group instructions somehow? 1. create the alg, 2. stretch the key 3. get the vector (or any other grouping that you like really :))

File: components/support/rc_crypto/nss/src/pbkdf2.rs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+// Expose for consumers to choose the hasing algorithm
+// Currently only SHA256 supported
+pub use crate::pk11::context::HashAlgorithm;
+use nss_sys::SECOidTag;
+use std::convert::TryFrom;
+
+// ***** BASED ON THE FOLLOWING IMPLEMENTATION *****
+// https://github.com/sqlcipher/sqlcipher/blob/4a81bea61e1da6fec222d713852830f1fd01aed2/src/crypto_nss.c#L182-L232
+// Which I believe is based on how it is done on WebCrypto:
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/pbkdf2.rs:19]: Nit: We can probably remove these and only keep the searchfox link

File: components/support/rc_crypto/nss/src/util.rs [hunk 15]
```diff
@@ -120,6 +120,43 @@ macro_rules! scoped_ptr {
             }
         }
     };
+    // For the case the destructor requires an addtional argument
```
Comment (eoger) [inline components/support/rc_crypto/nss/src/util.rs:123]: As mentioned previously I don't think you need this.

File: components/support/rc_crypto/src/pbkdf2.rs [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+/// let mut out = vec![0u8; 32];
+/// pbkdf2::pbkdf2_generate(&mut password, &mut salt, 2, pbkdf2::HashAlgorithm::SHA256, &mut out).unwrap(); // Oh oh should handle the error!
+/// assert_eq!(hex::encode(out), "ae4d0c95af6b46d32d0adff928f06dd02a303f8ef3c251dfd6e2d85a95474c43");
+//
+///```
+///
+/// # Erros
+///
+/// Could possibly return an error if the HMAC algorithm fails
+pub fn pbkdf2_generate(
```
Comment (eoger) [inline components/support/rc_crypto/src/pbkdf2.rs:32]:
```suggestion
pub fn derive(
```

(we already are in the `pbkdf2` module)

File: components/support/rc_crypto/src/pbkdf2.rs [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+/// use rc_crypto::pbkdf2;
+/// let mut password = b"password".to_vec();
+/// let mut salt = b"salt".to_vec();
+/// let mut out = vec![0u8; 32];
+/// pbkdf2::pbkdf2_generate(&mut password, &mut salt, 2, pbkdf2::HashAlgorithm::SHA256, &mut out).unwrap(); // Oh oh should handle the error!
+/// assert_eq!(hex::encode(out), "ae4d0c95af6b46d32d0adff928f06dd02a303f8ef3c251dfd6e2d85a95474c43");
+//
+///```
+///
+/// # Erros
```
Comment (eoger) [inline components/support/rc_crypto/src/pbkdf2.rs:29]:
```suggestion
/// # Errors
```

File: components/support/rc_crypto/nss/src/pbkdf2.rs (hunk 5 above)
Comment (tarikeshaq) [reply components/support/rc_crypto/nss/src/pbkdf2.rs]: I can't believe I didn't see that 🤦 Thanks @eoger and I really appreciate your reviews 🙏

File: components/support/rc_crypto/src/pbkdf2.rs [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    #[test]
+    fn test_more_iterations() {
+        let expected = "ae4d0c95af6b46d32d0adff928f06dd02a303f8ef3c251dfd6e2d85a95474c43";
+        let mut password = b"password".to_vec();
+        let mut salt = b"salt".to_vec();
+        let mut out = vec![0u8; 32];
+        derive(&mut password, &mut salt, 2, HashAlgorithm::SHA256, &mut out).unwrap();
+        assert_eq!(expected, hex::encode(out));
+    }
+}
```
Comment (rfk) [inline components/support/rc_crypto/src/pbkdf2.rs:75]: This seems to be working as expected, but I'd like to see more coverage of edge-cases in these tests, particularly around the things that might be going wrong when we transit from Rust into C. For example: do we properly handle embedded nulls in the password, and the salt? can we generate keys that aren't a full multiple of the hash blocksize? do we cleanly reject if requesting more than `(2^32 - 1) * hLen` bytes of output, assuming that NSS correctly fails in this case?

I think that `hkdf.rs` has a good sample of some of the things worth worrying about, but I also took the liberty of generating what I think would be some worthwhile extra test vectors, using the python `cryptography` library:

```
PBKDF2-SHA256(b'password', b'salt', 1, 32) = b'120fb6cffcf8b32c43e7225256c4f837a86548c92ccc35480805987cb70be17b'
PBKDF2-SHA256(b'password', b'salt', 1, 64) = b'120fb6cffcf8b32c43e7225256c4f837a86548c92ccc35480805987cb70be17b4dbf3a2f3dad3377264bb7b8e8330d4efc7451418617dabef683735361cdc18c'
PBKDF2-SHA256(b'password', b'salt', 2, 32) = b'ae4d0c95af6b46d32d0adff928f06dd02a303f8ef3c251dfd6e2d85a95474c43'
PBKDF2-SHA256(b'password', b'salt', 3, 33) = b'ad35240ac683febfaf3cd49d845473fbbbaa2437f5f82d5a415ae00ac76c6bfccf'
PBKDF2-SHA256(b'passw\x00rd', b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', 5, 13) = b'e25d526987819f966e324faa4a'
PBKDF2-SHA256(b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', b'salt', 2, 32) = b'62384466264daadc4144018c6bd864648272b34da8980d31521ffcce92ae003b'
PBKDF2-SHA256(b'password', b'salt', 1, 0) = b''
PBKDF2-SHA256(b'', b'salt', 1, 32) = b'f135c27993baf98773c5cdb40a5706ce6a345cde61b000a67858650cd6a324d7'
PBKDF2-SHA256(b'password', b'', 1, 32) = b'c1232f10f62715fda06ae7c0a2037ca19b33cf103b727ba56d870c11f290a2ab'
PBKDF2-SHA256(b'password', b'salt', 0, 32) = b'120fb6cffcf8b32c43e7225256c4f837a86548c92ccc35480805987cb70be17b'                                                                                                     
```

Comment (tarikeshaq) [reply components/support/rc_crypto/src/pbkdf2.rs:196]: Thank you for those!! I added the tests, but one thing I did notice, is that the NSS algorithm rejects 0 iterations and 0 length keyLength (out.len() ). I confirmed that via Crypto.subtle. But otherwise added the tests you mentioned and a few others to tests correct rejections.

Comment (rfk) [reply components/support/rc_crypto/src/pbkdf2.rs:196]: > NSS algorithm rejects 0 iterations and 0 length keyLength

Great :-)

TBH I was kind of surprised that the python implementation allowed these, but it's worth having tests to document the specific behaviour in these cases either way.

**Comment by tarikeshaq** [comment]
Thanks a lot, @rfk and @eoger !! I appreciate the reviews and feedback! 🕺
